ouroboros-consensus-cardano/tools/db-analyser/Analysis.hs

```diff
@@ -392,148 +426,188 @@ checkNoThunksEvery
-processAllUntil = either processAllImmutableDB processAllChainDB
-
-processAll ::
-     forall blk b st. (HasHeader blk, HasAnnTip blk)
-  => Either (ImmutableDB IO blk) (ChainDB IO blk)
-  -> ResourceRegistry IO
-  -> BlockComponent blk b
-  -> ExtLedgerState blk ValuesMK
-  -> Limit
-  -> st
-  -> (st -> b -> IO st)
-  -> IO st
-processAll db rr blockComponent initLedger limit initSt cb =
-  processAllUntil db rr blockComponent initLedger limit initSt callback
-    where
-      callback st b = (Continue, ) <$> cb st b
-
-processAll_ ::
-     forall blk b. (HasHeader blk, HasAnnTip blk)
-  => Either (ImmutableDB IO blk) (ChainDB IO blk)
-  -> ResourceRegistry IO
-  -> BlockComponent blk b
-  -> ExtLedgerState blk ValuesMK
-  -> Limit
-  -> (b -> IO ())
-  -> IO ()
-processAll_ db registry blockComponent initLedger limit callback =
-    processAll db registry blockComponent initLedger limit () (const callback)
-
-processAllChainDB ::
-     forall st blk b. (HasHeader blk, HasAnnTip blk)
-  => ChainDB IO blk
-  -> ResourceRegistry IO
-  -> BlockComponent blk b
-  -> ExtLedgerState blk ValuesMK
-  -> Limit
-  -> st
-  -> (st -> b -> IO (NextStep, st))
-  -> IO st
-processAllChainDB chainDB registry blockComponent (ExtLedgerState {..}) limit initState callback = do
-    itr <- case headerStateTip headerState of
-      Origin           -> ChainDB.streamAll
-                             chainDB
-                             registry
-                             blockComponent
-      NotOrigin annTip -> ChainDB.streamFrom
-                             (StreamFromExclusive $ annTipPoint annTip)
-                             chainDB
-                             registry
-                             blockComponent
-    go itr limit initState
+initialLedger ::
+  ( LedgerSupportsProtocol blk
+  , Serialise (HeaderHash blk)
+  , HasAnalysis blk
+  )
+  => SomeHasFS IO
+  -> TopLevelConfig blk
+  -> Either DiskSnapshot (ExtLedgerState blk ValuesMK)
+  -> IO (ExtLedgerState blk EmptyMK)
+initialLedger fs cfg e = do
+  r <- either
+    (\s -> either (error . show) id
+           <$> runExceptT (readSnapshot fs (decodeExtLedgerState' cfg) decode s))
+    (return . forgetLedgerTables) e
+  -- This marker divides the "loading" phase of the program, where the
+  -- system is principally occupied with reading snapshot data from
+  -- disk, from the "processing" phase, where we are streaming blocks
+  -- and running the ledger processing on them.
+  Debug.traceMarkerIO "SNAPSHOT_LOADED"
+  pure r
+
+decodeExtLedgerState' :: forall s blk .
+     HasAnalysis blk
+  => TopLevelConfig blk
+  -> Decoder s (ExtLedgerState blk EmptyMK)
+decodeExtLedgerState' cfg =
+  let ccfg = configCodec cfg
+  in decodeExtLedgerState
+           (decodeDisk ccfg)
+           (decodeDisk ccfg)
+           (decodeDisk ccfg)
+
+-- | This is quite similar to Ou*.OnDisk.replayStartingWith.push but there we
+-- cannot inject the check we are doing here with the callback. Also it performs
```

Comment:
To discuss: why can't we inject such check? Isn't this function more general than the one in `Ou*.OnDisk.replayStartingWith.push`?

Reply:
We could refactor `push` to accept `ApplyVal` or `ReapplyVal`. Not sure if we wanted to do this, as `ApplyVal` needs some handling of the ledger errors that can be thrown (`ThrowsLedgerErr`)